plugins/formal-verification/Uniqueness.md

```diff
@@ -174,10 +219,121 @@ class Stack(var root: @Unique Node?) {
 }
 ```
 
-[1]: https://dl.acm.org/doi/pdf/10.1145/130943.130947
+In this example, both `Node` and `Stack` can only
+meaningfully be used uniquely.
+One possibility is to require annotating the class itself
+with `Unique`, which would make the semantics clearer.
+However, it is not the only option: we could say that
+modifying the `root` property of a shared instance of
+`Stack` is permitted, but only using a `replace` operation
+that atomically sets it to new `@Unique Node?`.
 
-[2]: https://onlinelibrary.wiley.com/doi/abs/10.1002/spe.370
+The `value` property of `Node` is in this example forced to
+be unique, but in practice the user may want a `Stack` to be
+polymorphic in the uniqueness of its elements.
+It is an interesting question whether this can easily be
+supported.
 
-[3]: https://dl.acm.org/doi/pdf/10.1145/271510.271517
 
-[4]: https://arxiv.org/pdf/2309.05637.pdf
\ No newline at end of file
+## Benefits of Uniqueness
+
+Uniqueness annotations give the compiler information and
+that would otherwise be known only to the programmer.
+This also brings benefits to programmers who do not care
+about formal verification.
+
+### Smart casts
+
+In Kotlin, smart casts are defined based on a set of [stable
+expressions][6].
+Properties of objects referred to by a stable unique
+reference can themselves be considered stable.
+This allows us to accept the following code:
+
+```kt
+class X(var n: Int?)
+
+fun useX(x: @Borrowed @Unique X): Int =
+    if (x.n != null) {
+        x.n // Uniqueness grants that 'x.n' have not changed after checking is nullability
+    } else {
+        0
+    }
+```
+
+Note that the `@Unique` annotation permits the smart cast,
+while the `@Borrowed` annotation allows the caller of `useX`
+to not lose the uniqueness of the argument.
+
+### Analysis
+
+Aliasing in Kotlin also poses a challenge for the static
+analyzer used by IntelliJ IDEA.
+In the following example, aliasing results in an incorrect
+constant condition as provided by the IDE.
+
+```kt
+class A(var a: Boolean = false)
+
+fun f(a1: A, a2: A) {
+    a1.a = true
+    a2.a = false
+    if (!a1.a) {
+//      ^^^^^
+// Condition '!a1.a' is always false 
+        println("ALIASED!")
+    }
+}
+
+fun main() {
+    val a1 = A()
+    f(a1, a1) // prints "ALIASED!"
+}
+```
+
+### In-place modification
+
+When writing in a more functional style, operations on
+collections may look something like the following:
+
+```kt
+myList.map { ... }.filter { ... }
+```
+
+Both `map` and `filter` will create a new list to store the
+(intermediate) results.
+For the result of `map` in particular, we know that the list
+will only be used once, and so `filter` could operate
+in-place.
+
+By providing overloads on these functions that operate on
+unique and non-unique lists, we can reuse the list when
+we know that it cannot be observed later in the code.
+
```

Comment:
Perhaps it could be nice to have an example explaining better the advantage brought by uniqueness in this case. Because at first glance it was not too clear to me.

Reply:
I've tried to clarify, let me know if it makes more sense like this.  I think the example with the map and filter above is the most meaningful one, I'm not really sure what a second example would look like